### PR TITLE
fix(requires-writer): use `@reach/router` path ranking algorithm to sort matchPaths

### LINKED ISSUE #19043: matchPath priority takes over statically generated pages after version 2.15.22 and up to 2.17.4
## Description

I generate events with the `createPage` action using an `event.js` template. But not all events are pre-generated using this method, so I also have an `event.js` page that has a `matchPath = '/:year/:month/:day/:slug'` to fallback and dynamically display unknown events.
This was working great, up to Gatsby version 2.15.22. This is what runs at https://lebikini.com

For example this is pre-generated:
https://lebikini.com/2019/12/10/tiken-jah-fakoly/

And this is not:
https://lebikini.com/2019/10/03/deluxe-sora/

The idea is to not generate all past events as it is resource-intensive.

But with the recent modifications in Gatsby about page metadata, this stopped working in my case, so I locked the Gatsby version at 2.15.22.

Then I tried to upgrade again yesterday but even #18478 published in 2.17.3 does not fix it.

### Steps to reproduce

Minimum reproducible example using the default starter:
https://github.com/antoinerousseau/gatsby-bug/commit/671ae5e425d9e9fc8db61b30bec8ca7f22422b34

You can see it in action here:
https://gatsby-bug.netlify.com/2019/10/26/test-event/

### Expected result

We should see the static page created using the event template.

### Actual result

We see the fallback page because of its `matchPath`.

### Environment

```
  System:
    OS: macOS 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
    Yarn: 1.19.1 - ~/.nvm/versions/node/v10.15.3/bin/yarn
    npm: 6.12.0 - ~/.nvm/versions/node/v10.15.3/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 77.0.3865.120
    Firefox: 69.0
    Safari: 13.0.2
  npmPackages:
    gatsby: ^2.17.4 => 2.17.4
    gatsby-image: ^2.2.29 => 2.2.29
    gatsby-plugin-manifest: ^2.2.23 => 2.2.23
    gatsby-plugin-offline: ^3.0.16 => 3.0.16
    gatsby-plugin-react-helmet: ^3.1.13 => 3.1.13
    gatsby-plugin-sharp: ^2.2.32 => 2.2.32
    gatsby-source-filesystem: ^2.1.33 => 2.1.33
    gatsby-transformer-sharp: ^2.3.0 => 2.3.0
```

## Related issues

#18604 
#17980

### PR DESCRIPTION
**Pull request by pieh**
(this is draft, not tested outside of checking if `requires-writer.js` unit tests pass

TO-DO:
- [x] add test cases to e2e tests and not only unit tests, so behaviour is tested (and not just implementation).

I've added few "integration style" tests there:
 - `will find static path before dynamic paths`
 - `will find path with dynamic paramter before path with wildcard`
 - `it will find most specific path with wildcard`

Fixes #19043

### CODE REVIEW

File: packages/gatsby/src/bootstrap/__tests__/requires-writer.js [hunk 1]
```diff
@@ -148,6 +149,10 @@ describe(`requires-writer`, () => {
 
     it(`should have index pages with higher priority than matchPaths`, async () => {
       const pages = generatePagesState([
+        {
+          path: `/another-custom-404`,
+          matchPath: `/*`,
+        },
```
Comment (pieh) [inline packages/gatsby/src/bootstrap/__tests__/requires-writer.js]: (to make sure order of created pages doesn't impact final order of sorted pages)

File: packages/gatsby/src/bootstrap/requires-writer.js [hunk 2]
```diff
@@ -80,20 +102,15 @@ const getMatchPaths = pages => {
...
       const order = b.score - a.score
       if (order !== 0) {
         return order
       }
 
-      // if specificity is the same we use the array index
-      return b.index - a.index
+      // if specificity is the same we do lexigraphic comparison of path to ensure
+      // deterministic order regardless of order pages where created
+      return a.matchPath.localeCompare(b.matchPath)
```
Comment (pieh) [inline packages/gatsby/src/bootstrap/requires-writer.js]: This also need some eyes on (this is not exactly something needed for this fix, it just should mean that `matchPaths.json` order is always the same regardless of order of pages)

**Comment by sever1an** [comment]
Is this the function we would ask reach-router to export? Should we open an issue there?
https://github.com/reach/router/blob/28a79e7fc3a3487cb3304210dc3501efb8a50eba/src/lib/utils.js#L228

**Comment by pieh** [comment]
> Is this the function we would ask reach-router to export? Should we open an issue there?
> [reach/router:src/lib/utils.js@`28a79e7`#L228](https://github.com/reach/router/blob/28a79e7fc3a3487cb3304210dc3501efb8a50eba/src/lib/utils.js#L228)

I did modify the function a little bit (I'm passing just path instead of route object to avoid creating temporary objects - micro-optim), but essentially yes.

In my mind this shouldn't be a blocker for this PR - so this is opportunity for potential follow-up PRs (we don't need to hold off fix and wait for `@reach/router` to release new version - latest stable release was ~1 year ago, and they have committed more changes recently that are under `next` tag, so it might be a while for next stable version release)

**Comment by sever1an** [comment]
I agree, it should be a followup pr/issue. We are already hosting a parallel implementation of route scoring, so there is no extra maintenance. 

Kind of thinking if we want it to happen before Gatsby 3.0.0, maybe start sooner than later....

**Comment by pieh** [comment]
Added some e2e tests together with @madalynrose:

For reference:
e2e tests that I added have following results on:
 - current `latest`:
   Screenshot 2019-10-31 at 20 14 21
 - `gatsby@2.17.2` (before wildcard fix):
   Screenshot 2019-10-31 at 20 16 31

File: yarn.lock [hunk 3]
```diff
@@ -6926,7 +6926,7 @@ csstype@^2.2.0, csstype@^2.5.7:
   resolved "https://registry.yarnpkg.com/csstype/-/csstype-2.6.5.tgz#1cd1dff742ebf4d7c991470ae71e12bb6751e034"
   integrity sha512-JsTaiksRsel5n7XwqPAfB0l3TFKdpjW/kgAELf9vrb5adGA7UCPLajKK5s3nFrcFm3Rkyp/Qkgl73ENc1UY3cA==
 
-"csstype@github:rtsao/csstype#c4e709f":
+csstype@rtsao/csstype#c4e709f:
```
Comment (wardpeet) [inline yarn.lock]: I don't see any package.json changes,not saying this is bad. it's a little bit weird that's all :) (not a blocker)

**Comment by pieh** [comment]
Published `gatsby@2.17.9`

**Pull request merged**